clustered/server/src/main/java/org/ehcache/clustered/server/store/ClusterTierActiveEntity.java

```diff
@@ -586,40 +594,74 @@ public void synchronizeKeyToPassive(PassiveSynchronizationChannel<EhcacheEntityM
     if (concurrencyKey == DEFAULT_KEY) {
       stateService.getStateRepositoryManager().syncMessageFor(storeIdentifier).forEach(syncChannel::synchronizeToPassive);
     } else {
+      BlockingQueue<EhcacheDataSyncMessage> messageQ = new SynchronousQueue<>();
       int segmentId = concurrencyKey - DEFAULT_KEY - 1;
-      Long dataSizeThreshold = Long.getLong(SYNC_DATA_SIZE_PROP, DEFAULT_SYNC_DATA_SIZE_THRESHOLD);
-      AtomicLong size = new AtomicLong(0);
-      ServerSideServerStore store = stateService.getStore(storeIdentifier);
-      final AtomicReference<Map<Long, Chain>> mappingsToSend = new AtomicReference<>(new HashMap<>());
-      store.getSegmentKeySets().get(segmentId)
-        .forEach(key -> {
-          final Chain chain;
-          try {
-            chain = store.get(key);
-          } catch (TimeoutException e) {
-            throw new AssertionError("Server side store is not expected to throw timeout exception");
-          }
-          for (Element element : chain) {
-            size.addAndGet(element.getPayload().remaining());
-          }
-          mappingsToSend.get().put(key, chain);
-          if (size.get() > dataSizeThreshold) {
-            syncChannel.synchronizeToPassive(new EhcacheDataSyncMessage(mappingsToSend.get()));
-            mappingsToSend.set(new HashMap<>());
-            size.set(0);
+      CompletableFuture<Void> asyncGet = CompletableFuture.runAsync(() -> doGetsForSync(segmentId, messageQ));
+      try {
+        boolean done = false;
+        while (!done) {
+          EhcacheDataSyncMessage msg = messageQ.take();
+          if (msg instanceof LastDataSyncMessage) {
+            done = true;
+          } else {
+            syncChannel.synchronizeToPassive(msg);
           }
-        });
-      if (!mappingsToSend.get().isEmpty()) {
-        syncChannel.synchronizeToPassive(new EhcacheDataSyncMessage(mappingsToSend.get()));
-        mappingsToSend.set(new HashMap<>());
-        size.set(0);
+        }
+        asyncGet.get();
+      } catch (InterruptedException | ExecutionException e) {
+        throw new RuntimeException(e);
```

Comment:
Not a fan of the interruption management. I would interrupt the loop instead, and tell the other thread so stop the task ideally.

Reply:
Using interrupts to signal here seems fragile to me too.  I'm thinking we could maybe restructure this asas a SynchronousQueue of functional types.  Then we can push all the sync logic on to one side of the queue and have the consumer just be:

```java
while(synchronousQueue.take().execute());
```